warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java

```diff
@@ -982,6 +982,161 @@ private static Map<String,Object> getAssignments(JexlNode node, Map<String,Objec
+    }
+
+    /**
+     * This method accepts a single replacement node (leaf node) and attempts to merge it cleanly into the parent or grandparent while avoiding a reference
+     * expression as the new parent.
+     *
+     * @param parent
+     *            the parent node
+     * @param original
+     *            the original node
+     * @param replacement
+     *            the replacement node
+     */
+    private static void replaceNodeWithSingleValue(JexlNode parent, JexlNode original, JexlNode replacement) {
+
+        if (!(parent instanceof ASTReferenceExpression)) {
+            // straight replacement
+            JexlNodes.replaceChild(parent, original, replacement);
+            return;
+        }
+
+        // otherwise we have a reference expression and we need to go up a layer
+        JexlNode grandParent = parent.jjtGetParent();
+
+        if (grandParent != null) {
+            // check for the negated single term case: !(term)
+            // it is also unlikely that two reference expressions would sit next to each other in the tree, but it's
+            // not impossible
+            if (!(grandParent instanceof ASTNotNode) && !(grandParent instanceof ASTReferenceExpression)) {
+                JexlNodes.replaceChild(grandParent, parent, replacement);
+                return;
+            }
+        }
+
+        // if no optimal rewrite occurred fallback to a simple replacement
+        JexlNodes.replaceChild(parent, original, replacement);
+    }
+
+    /**
+     * Determines if the provided JexlNode is a junction
+     *
+     * @param node
+     *            the JexlNode
+     * @return true if the node is a junction
+     */
+    public static boolean isJunction(JexlNode node) {
+        return !QueryPropertyMarker.findInstance(node).isAnyType() && (node instanceof ASTOrNode || node instanceof ASTAndNode);
+    }
+
+    /**
+     * Determines if the two JexlNodes are similar junction types (both And nodes or both Or nodes)
+     *
+     * @param a
+     *            first JexlNode
+     * @param b
+     *            second JexlNode
+     * @return true if the two nodes are equivalent junctions
+     */
+    public static boolean isEquivalentJunctions(JexlNode a, JexlNode b) {
+        if (QueryPropertyMarker.findInstance(a).isAnyType() || QueryPropertyMarker.findInstance(b).isAnyType()) {
+            return false;
+        }
+
+        return (a instanceof ASTOrNode && b instanceof ASTOrNode) || (a instanceof ASTAndNode && b instanceof ASTAndNode);
+    }
+
+    /**
+     * Given two equivalent junctions per {@link #isEquivalentJunctions(JexlNode, JexlNode)}, merges the children for node B into the children of node A
+     *
+     * @param parent
+     *            the parent
+     * @param a
+     *            first JexlNode
+     * @param b
+     *            second JexlNode
+     */
+    public static void mergeJunctions(JexlNode parent, JexlNode a, JexlNode b) {
```

Comment:
This has some preconditions that need to be checked. No guarantees that parent is related to a or b

Reply:
this is a public static method you can't rely on the calling method to validate the internals